llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -525,26 +646,144 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
   AspectsSetTy LocalAspects;
   for (Function *Callee : It->second) {
     if (Visited.insert(Callee).second)
-      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+      propagateAspectsThroughCG(Callee, CG, AspectsMap,
+                                FunctionsCalledUnconditionally, Visited);
 
     const auto &CalleeAspects = AspectsMap[Callee];
     LocalAspects.insert(CalleeAspects.begin(), CalleeAspects.end());
   }
+  if (FunctionsCalledUnconditionally.contains(F))
+    AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+using PathsContainingConditionalCallersTy =
+    std::vector<std::vector<Function *>>;
+
+void identifyPathsContainingConditionalCallers(
+    Function *F, const CallGraphTy &CG, std::vector<Function *> &CurrentPath,
+    PathsContainingConditionalCallersTy &Paths) {
+  const auto It = CG.find(F);
+  if (It == CG.end()) {
+    CurrentPath.push_back(F);
+    for (auto Func : CurrentPath)
+      if (Func->hasFnAttribute("sycl-call-if-on-device-conditionally")) {
+        Paths.push_back(CurrentPath);
+        break;
+      }
+    CurrentPath.pop_back();
+    return;
+  }
+
+  CurrentPath.push_back(F);
+
+  for (Function *Callee : It->second) {
+    identifyPathsContainingConditionalCallers(Callee, CG, CurrentPath, Paths);
+  }
 
-  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+  CurrentPath.pop_back();
+}
+
+void propagateConditionExpressionsAndConditionalAspectsThroughCG(
+    const PathsContainingConditionalCallersTy &Paths,
+    FunctionToConditionExpressionsAndConditionalAspectsTy
+        &ConditionsAndAspectsMap,
+    FunctionToConditionExpressionsMapTy &ConditionsMap,
+    FunctionToAspectsMapTy &AspectsMap) {
+  auto Contains =
+      [](ConditionExpressionsAndConditionalAspectsTy ConditionsAndAspectsVec,
+         std::pair<ConditionsSetTy, AspectsSetTy> Search) {
+        for (const auto &ConditionAndAspectsPair : ConditionsAndAspectsVec)
+          if ((ConditionAndAspectsPair.first == Search.first) &&
+              (ConditionAndAspectsPair.second == Search.second))
+            return true;
+        return false;
+      };
+
+  // TODO: need to optimize finding duplicates and combining pairs with the same
+  // Condition Expressions or Aspects
+  for (const auto &Path : Paths)
+    for (int I = Path.size() - 1; I >= 0; --I) {
+      if (I != Path.size() - 1) {
```

Comment:
```suggestion
    for (int E = Path.size() - 1, I = E; I >= 0; --I) {
      if (I != E) {
```

I'm not sure if `I = E` is possible there, but if not, you can repeat `Path.size() - 1` for `I` initialization, that would anyway allow you to avoid repeating it within the loop by re-using `E` instead and reducing potential amount of off-by-one errors